src/Compilers/Core/Portable/Collections/ConcurrentCache.cs

```diff
@@ -37,7 +30,11 @@ internal Entry(int hash, TKey key, TValue value)
         }
 
         public ConcurrentCache(int size, IEqualityComparer<TKey> keyComparer)
-            : base(size)
+            // Defer creating the backing array until it is actually needed.  This saves on expensive allocations for
+            // short-lived compilations that do not end up using the cache.  As the cache is simple best-effort, it's
+            // fine if multiple threads end up creating the backing array at the same time.  One thread will be last and
+            // will win, and the others will just end up creating a small piece of garbage that will be collected.
+            : base(size, createBackingArray: false)
```

Comment:
i kept this logic *only* for ConcurrentCache, as it's this cache that says explicitly: `expiration policy is "new entry wins over old entry if hashed into the same bucket"`.  So it felt safest to have just for this type.  We could move this down to the base if we're ok with this policy for all subclsses.